pkg/snmp/traps/config.go

```diff
@@ -60,12 +82,90 @@ func (c *Config) Addr() string {
 	return fmt.Sprintf("%s:%d", c.BindHost, c.Port)
 }
 
-// BuildV2Params returns a valid GoSNMP SNMPv2 params structure from configuration.
-func (c *Config) BuildV2Params() *gosnmp.GoSNMP {
-	return &gosnmp.GoSNMP{
-		Port:      c.Port,
-		Transport: "udp",
-		Version:   gosnmp.Version2c,
-		Logger:    gosnmp.NewLogger(&trapLogger{}),
+func (c *Config) getAgentHostname() string {
+	hostname, err := util.GetHostname(context.TODO())
+	if err != nil {
+		// this scenario is not likely to happen since
+		// the agent cannot start without a hostname
+		hostname = "unknown-datadog-agent"
 	}
+	return hostname
+}
+
+// BuildAuthoritativeEngineID creates a engineID byte-array from the OS hostname
+func (c *Config) BuildAuthoritativeEngineID() [28]byte {
+	engineID := [28]byte{}
+	// First byte is always 0x80
+	// Next four bytes are the Private Enterprise Number (set to an invalid value here)
+	copy(engineID[:5], []byte{0x80, 0xff, 0xff, 0xff, 0xff})
+	copy(engineID[5:], []byte(c.getAgentHostname()))
+	return engineID
+}
+
+// BuildSNMPParams returns a valid GoSNMP params structure from configuration.
+func (c *Config) BuildSNMPParams() (*gosnmp.GoSNMP, error) {
+	if len(c.Users) == 0 {
+		return &gosnmp.GoSNMP{
+			Port:      c.Port,
+			Transport: "udp",
+			Version:   gosnmp.Version2c, // No user configured, let's use Version2 which is enough and doesn't require setting up fake security data.
+			Logger:    gosnmp.NewLogger(&trapLogger{}),
+		}, nil
+	}
+	user := c.Users[0]
+	var authProtocol gosnmp.SnmpV3AuthProtocol
+	lowerAuthProtocol := strings.ToLower(user.AuthProtocol)
+	if lowerAuthProtocol == "" {
+		authProtocol = gosnmp.NoAuth
+	} else if lowerAuthProtocol == "md5" {
+		authProtocol = gosnmp.MD5
+	} else if lowerAuthProtocol == "sha" {
+		authProtocol = gosnmp.SHA
+	} else {
+		return nil, fmt.Errorf("unsupported authentication protocol: %s", user.AuthProtocol)
+	}
+
+	var privProtocol gosnmp.SnmpV3PrivProtocol
+	lowerPrivProtocol := strings.ToLower(user.PrivProtocol)
+	if lowerPrivProtocol == "" {
+		privProtocol = gosnmp.NoPriv
+	} else if lowerPrivProtocol == "des" {
+		privProtocol = gosnmp.DES
+	} else if lowerPrivProtocol == "aes" {
+		privProtocol = gosnmp.AES
+	} else if lowerPrivProtocol == "aes192" {
+		privProtocol = gosnmp.AES192
+	} else if lowerPrivProtocol == "aes192c" {
+		privProtocol = gosnmp.AES192C
+	} else if lowerPrivProtocol == "aes256" {
+		privProtocol = gosnmp.AES256
+	} else if lowerPrivProtocol == "aes256c" {
+		privProtocol = gosnmp.AES256C
+	} else {
+		return nil, fmt.Errorf("unsupported privacy protocol: %s", user.PrivProtocol)
+	}
+
+	msgFlags := gosnmp.NoAuthNoPriv
+	if user.PrivKey != "" {
+		msgFlags = gosnmp.AuthPriv
+	} else if user.AuthKey != "" {
+		msgFlags = gosnmp.AuthNoPriv
+	}
```

Comment:
Similar code exist for corechecks/snmp and snmp_listener to convert authProtocol/privProtocol, shall we reuse the same code base ?

https://github.com/DataDog/datadog-agent/blob/main/pkg/collector/corechecks/snmp/gosnmplib/gosnmp_auth.go

Reply:
Yes, can be part of `II-799` as we plan to move `gosnmplib` to a shared pkg.